test/ir_Metz_test.cpp

```diff
@@ -0,0 +1,95 @@
+// Copyright 2020 crankyoldgit
+
+#include "IRac.h"
+#include "IRrecv.h"
+#include "IRrecv_test.h"
+#include "IRsend.h"
+#include "IRsend_test.h"
+#include "gtest/gtest.h"
+
+// Tests for decodeMetz().
+
+TEST(TestDecodeMetz, RealExample) {
+  IRsendTest irsend(kGpioUnused);
+  IRrecv irrecv(kGpioUnused);
+  uint16_t rawbuf[41] = {
+      // Header          1         0          1
+      880, 2336, 462, 1656, 460, 952, 486, 1630,
+      //      1          1         0         0
+      462, 1654, 460, 1660, 462, 948, 488, 902,
+      //      1         0          1         0
+      486, 1656, 486, 926, 486, 1630, 486, 924,
+      //      1         0         0          1
+      488, 1630, 460, 954, 460, 950, 512, 1606,
+      //     0          1         0          1
+      460, 978, 436, 1656, 486, 926, 486, 1630,
+      458};
+  irsend.begin();
+  irsend.reset();
+  irsend.sendRaw(rawbuf, 41, 38);
+  irsend.makeDecodeResult();
+
+  ASSERT_TRUE(irrecv.decode(&irsend.capture));
+  ASSERT_EQ(decode_type_t::METZ, irsend.capture.decode_type);
+  ASSERT_EQ(kMetzBits, irsend.capture.bits);
+  EXPECT_EQ(0x5CA95, irsend.capture.value);
+  EXPECT_EQ(0x2A, irsend.capture.address);
+  EXPECT_EQ(0x3, irsend.capture.command);
+}
+
+TEST(TestDecodeMetz, SyntheticExample) {
+  IRsendTest irsend(kGpioUnused);
+  IRrecv irrecv(kGpioUnused);
+  irsend.begin();
+  irsend.reset();
+  irsend.sendMetz(0x5CA95);
+  irsend.makeDecodeResult();
+
+  ASSERT_TRUE(irrecv.decode(&irsend.capture));
+  EXPECT_EQ(decode_type_t::METZ, irsend.capture.decode_type);
+  EXPECT_EQ(kMetzBits, irsend.capture.bits);
+  EXPECT_EQ(0x5CA95, irsend.capture.value);
+  EXPECT_EQ(0x2A, irsend.capture.address);
+  EXPECT_EQ(0x3, irsend.capture.command);
```

Comment:
The meanings of _command_ and _address_ seem to be swapped. Three bits for command would be quite low.

For the codes of my RM17, I created a table containing the output of Tasmota:

```python
>>> sorted([((int(v["IrReceived"]["Data"], base=0) >> 6) & 0x3f) for v in table.values()])
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46]
>>> sorted([((int(v["IrReceived"]["Data"], base=0) >> 15) & 0x7) for v in table.values()])
[3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]

```

Reply:
This might be better as a new issue.
Number of bits might be swaped https://github.com/ukw100/IRMP/blob/f8c21a5ae9fae9915fa3a5c37e75555fb870a393/src/irmpprotocols.h#L1040
https://github.com/crankyoldgit/IRremoteESP8266/pull/1242/files#diff-2921a99f02331f7b182350fdb3faa9aaR21-R24

But do we have some official documentation for the protocol instead of only implementations?
Maybe it would help if we had a set of 1-9 or similar keypresses? (if the codes are sequential)

Reply:
No, the number of bits is correct. Just the names of the fields are wrong.

I don't think this justifies a new issue. This PR hasn't been merged yet and it seems to be trivial to me.

There's no public documentation, as far as I know.

I compiled a list of all codes for my remotes:
```yaml
0:              {Bits: 19, Code: 0, Data: '0x1C03F', Protocol: METZ}
1:              {Bits: 19, Code: 1, Data: '0x1C07E', Protocol: METZ}
2:              {Bits: 19, Code: 2, Data: '0x1C0BD', Protocol: METZ}
3:              {Bits: 19, Code: 3, Data: '0x1C0FC', Protocol: METZ}
4:              {Bits: 19, Code: 4, Data: '0x1C13B', Protocol: METZ}
5:              {Bits: 19, Code: 5, Data: '0x1C17A', Protocol: METZ}
6:              {Bits: 19, Code: 6, Data: '0x1C1B9', Protocol: METZ}
7:              {Bits: 19, Code: 7, Data: '0x1C1F8', Protocol: METZ}
8:              {Bits: 19, Code: 8, Data: '0x1C237', Protocol: METZ}
9:              {Bits: 19, Code: 9, Data: '0x1C276', Protocol: METZ}
Power:          {Bits: 19, Code: 10, Data: '0x1C2B5', Protocol: METZ}
F1:             {Bits: 19, Code: 11, Data: '0x1C2F4', Protocol: METZ}
F2:             {Bits: 19, Code: 12, Data: '0x1C333', Protocol: METZ}
F3:             {Bits: 19, Code: 13, Data: '0x1C372', Protocol: METZ}
TV_R:           {Bits: 19, Code: 14, Data: '0x1C3B1', Protocol: METZ}
Blue_Video:     {Bits: 19, Code: 17, Data: '0x1C46E', Protocol: METZ}
Yellow_Audio:   {Bits: 19, Code: 18, Data: '0x1C4AD', Protocol: METZ}
Signal:         {Bits: 19, Code: 19, Data: '0x1C4EC', Protocol: METZ}
Green_Clock:    {Bits: 19, Code: 20, Data: '0x1C52B', Protocol: METZ}
Red_Asterisk:   {Bits: 19, Code: 21, Data: '0x1C56A', Protocol: METZ}
Up:             {Bits: 19, Code: 22, Data: '0x1C5A9', Protocol: METZ}
Down:           {Bits: 19, Code: 23, Data: '0x1C5E8', Protocol: METZ}
Left:           {Bits: 19, Code: 24, Data: '0x1C627', Protocol: METZ}
Right:          {Bits: 19, Code: 25, Data: '0x1C666', Protocol: METZ}
Vol_Plus:       {Bits: 19, Code: 26, Data: '0x1C6A5', Protocol: METZ}
Vol_Minus:      {Bits: 19, Code: 27, Data: '0x1C6E4', Protocol: METZ}
INFO:           {Bits: 19, Code: 28, Data: '0x1C723', Protocol: METZ}
PIP:            {Bits: 19, Code: 29, Data: '0x1C762', Protocol: METZ}
Return:         {Bits: 19, Code: 32, Data: '0x1C81F', Protocol: METZ}
Pause:          {Bits: 19, Code: 33, Data: '0x1C85E', Protocol: METZ}
Mute:           {Bits: 19, Code: 34, Data: '0x1C89D', Protocol: METZ}
Text:           {Bits: 19, Code: 35, Data: '0x1C8DC', Protocol: METZ}
EPG:            {Bits: 19, Code: 36, Data: '0x1C91B', Protocol: METZ}
MENU:           {Bits: 19, Code: 37, Data: '0x1C95A', Protocol: METZ}
OK:             {Bits: 19, Code: 38, Data: '0x1C999', Protocol: METZ}
Exit:           {Bits: 19, Code: 39, Data: '0x1C9D8', Protocol: METZ}
Fast_Rewind:    {Bits: 19, Code: 40, Data: '0x1CA17', Protocol: METZ}
Fast_Forward:   {Bits: 19, Code: 41, Data: '0x1CA56', Protocol: METZ}
Play:           {Bits: 19, Code: 42, Data: '0x1CA95', Protocol: METZ}
Stop:           {Bits: 19, Code: 43, Data: '0x1CAD4', Protocol: METZ}
Record:         {Bits: 19, Code: 44, Data: '0x1CB13', Protocol: METZ}
P_Plus:         {Bits: 19, Code: 45, Data: '0x1CB52', Protocol: METZ}
P_Minus:        {Bits: 19, Code: 46, Data: '0x1CB91', Protocol: METZ}
```
And I was able to find four more codes with Tasmota:
```yaml
AV:             {Bits: 19, Code: 15, Data: '0x1C3F0', Protocol: METZ}
Radio:          {Bits: 19, Code: 16, Data: '0x1C42F', Protocol: METZ}
Discrete_Off:   {Bits: 19, Code, 50, Data: '0x1CC8D', Protocol: METZ}
Discrete_On:    {Bits: 19, Code, 51, Data: '0x1CCCC', Protocol: METZ}
```

I created the "unknown" codes with this function:
```python
def metz(code, addr=3):
    return addr << 15 | (~addr & 7) << 12 | code << 6 | ~code & 0x3f
```

Reply:
My bad on the status and new issue, confused it with another PR.

Re-reading I agree that swapping all instances of address with command seems correct.
(I limited my view to just the bit sizes)
```c
const uint8_t kMetzAddressBits = 6;
const uint8_t kMetzCommandBits = 3;
```

Just to be sure in your list, Data is from this branch when using the remote? and the Code is with your function? (just trying to make sure all data is available without misunderstandings for @crankyoldgit to use.)

Reply:
Yes, Data in the bigger list is from this branch using the remote. Code is just (Data >> 6) & 0x3f.

In the smaller list, Data was created with Code as input to my function.

Reply:
@mtdcr I think I've successfully swapped them around in the latest commit to this branch.
Can you please confirm the results are as you expect now?

Reply:
@mtdcr ping.